pubsub/src/main/java/com/example/pubsub/PubSubExample.kt

```diff
@@ -0,0 +1,189 @@
+    }
+
+    // Your topic ID, eg. "my-topic"
+    val topicId = args[0]
+
+    // Create a new topic
+    val topic = ProjectTopicName.of(projectId, topicId)
+
+    try {
+        TopicAdminClient.create().use { topicAdminClient -> topicAdminClient.createTopic(topic) }
+        println("Topic ${topic.project}:${topic.topic} successfully created.")
+    } catch (e: ApiException) {
+        // example : code = ALREADY_EXISTS(409) implies topic already exists
+        print(e.statusCode.code)
+        print(e.isRetryable)
+    }
+}
+
+private fun subscribeTopic(vararg args: String) { // expects 2 args: <topic> and <subscription>
+
+    if (args.size < 2) {
+        error("Bad input: command 'sub' expects 2 arguments. \n $usage")
+    }
+
+    // Your topic ID, eg. "my-topic"
+    val topicId = args[0]
+
+    // Your subscription ID eg. "my-sub"
+    val subscriptionId = args[1]
+
+    val topicName = ProjectTopicName.of(projectId, topicId)
+
+    // Create a new subscription
+    val subscriptionName = ProjectSubscriptionName.of(projectId, subscriptionId)
+
+    try {
+        SubscriptionAdminClient.create().use { subscriptionAdminClient ->
+            // create a pull subscription with default acknowledgement deadline (= 10 seconds)
+            subscriptionAdminClient.createSubscription(subscriptionName, topicName, PushConfig.getDefaultInstance(), 0)
+        }
+        println("Subscription ${subscriptionName.project}:${subscriptionName.subscription} successfully created.")
+    } catch (e: ApiException) {
+        // example : code = ALREADY_EXISTS(409) implies subscription already exists
+        print(e.statusCode.code)
+        print(e.isRetryable)
+    }
+}
+
+private fun publishMsg(vararg args: String) { // expects 2 args: <topic> and <count>
+
+    if (args.size < 2) {
+        error("Bad input: command 'sub' expects 2 arguments. \n $usage")
+    }
+    val topicId = args[0]
+    val messageCount = args[1].toInt()
+    val topicName = ProjectTopicName.of(projectId, topicId)
+    lateinit var publisher: Publisher
+    val futures: MutableList<ApiFuture<String>> = mutableListOf()
+
+    try {
+        // Create a publisher instance with default settings bound to the topic
+        publisher = Publisher.newBuilder(topicName).build()
+
+        for (i in 1..messageCount) {
+            val message = "message-$i"
+
+            // convert message to bytes
+            val data = ByteString.copyFromUtf8(message)
+            val pubsubMessage = PubsubMessage.newBuilder()
+                    .setData(data)
+                    .build()
+
+            // Schedule a message to be published. Messages are automatically batched.
+            val future = publisher.publish(pubsubMessage)
+            futures.add(future)
+        }
+    } finally {
+        // Wait on any pending requests
+        val messageIds = ApiFutures.allAsList(futures).get()
+        messageIds.forEach { messageId -> println(messageId) }
+        publisher.shutdown()
```

Comment:
It isn't clear in the sample what the IDs are being printed. I would suggest printing `"Published the following messages: "` and then listing them as `"$messageId - $message"`, or something along those lines

Reply:
Not easy stuff. publisher.publish returns a "future". On success, it returns only the id of the message published. This is all we have. The only way to have messages data along w/ their ids is to "listen" to a sub.

Reply:
Ahh I see, yes we don't want to do that. How about we add `println("Published Message IDs:")` then, above the list you already have.